experimental/modular_oak_functions/README.md

```diff
@@ -0,0 +1,75 @@
+# Proposal for making Oak Functions Extensions more general
+
+We currently have ExtensionFactories that generate Extensions. Extensions
+provide the functions available to the per-request Wasm business logic
+instances. This document proposes a generalisation of extensions and the ideas
+mentioned in issue [#2551](https://github.com/project-oak/oak/issues/2551).
+
+## Naming
+
+We have previously discussed using a more general term, such as "Service" rather
+than "Extension". My suggestion is to use `Service` as a replacement for
+`ExtensionFactory` and `ServiceProxy` as a replacement for `Extension`.
+
+The shared state used by an extension is generally owned and managed by the
+extension factory, and each extension is specific to a single `WasmState`
+instance that manages the Wasm execution for a single request. I also considered
+`ServiceFactory` and `Service` as respective replacements, but seeing that the
+`ServiceFactory` would also manage shared state it seems somewhat inaccurate.
+
+## Design Approach
+
+The `Service` implementation for a feature would manage shared state and create
+`ServiceProxy` instances that can be used to perform per-request service-related
+actions, and interact with the shared state owned by its parent service where
+needed.
+
+Each service implementation would decide whether it must be able to call other
+services, and if so provide the ability to inject these services into a
+constructor. Each service will also be able to receive configuration information
+through a standardised interface.
```

Comment:
It could be arbitrary, but in our case it is mostly linear. It is only the Wasm service that allows the business logic to choose different services to send data to.